File: kreuzberg/packages/java/src/main/java/dev/kreuzberg/DjotContent.java
package dev.kreuzberg;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * Comprehensive Djot document structure with semantic preservation.
 *
 * <p>
 * Captures the full richness of Djot markup, including:
 * <ul>
 * <li>Block-level structures (headings, lists, blockquotes, code blocks,
 * etc.)</li>
 * <li>Inline formatting (emphasis, strong, highlight, subscript, superscript,
 * etc.)</li>
 * <li>Attributes (classes, IDs, key-value pairs)</li>
 * <li>Links, images, footnotes</li>
 * <li>Math expressions (inline and display)</li>
 * <li>Tables with full structure</li>
 * </ul>
 *
 * <p>
 * Available when the Djot feature is enabled.
 *
 * @since 0.8.0
 */
public final class DjotContent {
	private final String plainText;
	@JsonDeserialize(contentAs = FormattedBlock.class)
	private final List<FormattedBlock> blocks;
	private final Metadata metadata;
	@JsonDeserialize(contentAs = Table.class)
	private final List<Table> tables;
	@JsonDeserialize(contentAs = DjotImage.class)
	private final List<DjotImage> images;
	@JsonDeserialize(contentAs = DjotLink.class)
	private final List<DjotLink> links;
	@JsonDeserialize(contentAs = Footnote.class)
	private final List<Footnote> footnotes;
	@JsonDeserialize(contentAs = AttributeEntry.class)
	private final List<AttributeEntry> attributes;

	@JsonCreator
	public DjotContent(@JsonProperty("plain_text") String plainText,
			@JsonProperty("blocks") List<FormattedBlock> blocks, @JsonProperty("metadata") Metadata metadata,
			@JsonProperty("tables") List<Table> tables, @JsonProperty("images") List<DjotImage> images,
			@JsonProperty("links") List<DjotLink> links, @JsonProperty("footnotes") List<Footnote> footnotes,
			@JsonProperty("attributes") List<AttributeEntry> attributes) {
		this.plainText = Objects.requireNonNull(plainText, "plainText must not be null");
		this.blocks = Collections.unmodifiableList(blocks != null ? new ArrayList<>(blocks) : new ArrayList<>());
		this.metadata = metadata != null ? metadata : Metadata.empty();
		this.tables = Collections.unmodifiableList(tables != null ? new ArrayList<>(tables) : new ArrayList<>());
		this.images = Collections.unmodifiableList(images != null ? new ArrayList<>(images) : new ArrayList<>());
		this.links = Collections.unmodifiableList(links != null ? new ArrayList<>(links) : new ArrayList<>());
		this.footnotes = Collections
				.unmodifiableList(footnotes != null ? new ArrayList<>(footnotes) : new ArrayList<>());
		this.attributes = Collections
				.unmodifiableList(attributes != null ? new ArrayList<>(attributes) : new ArrayList<>());
	}

	/**
	 * Get the plain text representation of the document.
	 *
	 * @return plain text
	 */
	public String getPlainText() {
		return plainText;
	}

	/**
	 * Get the block-level content of the document.
	 *
	 * @return unmodifiable list of blocks (never null)
	 */
	public List<FormattedBlock> getBlocks() {
		return blocks;
	}

	/**
	 * Get the document metadata.
	 *
	 * @return metadata (never null)
	 */
	public Metadata getMetadata() {
		return metadata;
	}

	/**
	 * Get the extracted tables.
	 *
	 * @return unmodifiable list of tables (never null)
	 */
	public List<Table> getTables() {
		return tables;
	}

	/**
	 * Get the extracted images.
	 *
	 * @return unmodifiable list of images (never null)
	 */
	public List<DjotImage> getImages() {
		return images;
	}

	/**
	 * Get the extracted links.
	 *
	 * @return unmodifiable list of links (never null)
	 */
	public List<DjotLink> getLinks() {
		return links;
	}

	/**
	 * Get the footnote definitions.
	 *
	 * @return unmodifiable list of footnotes (never null)
	 */
	public List<Footnote> getFootnotes() {
		return footnotes;
	}

	/**
	 * Get the element attributes mapped by identifier.
	 *
	 * @return unmodifiable list of attribute entries (never null)
	 */
	public List<AttributeEntry> getAttributes() {
		return attributes;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (!(obj instanceof DjotContent)) {
			return false;
		}
		DjotContent other = (DjotContent) obj;
		return Objects.equals(plainText, other.plainText) && Objects.equals(blocks, other.blocks)
				&& Objects.equals(metadata, other.metadata) && Objects.equals(tables, other.tables)
				&& Objects.equals(images, other.images) && Objects.equals(links, other.links)
				&& Objects.equals(footnotes, other.footnotes) && Objects.equals(attributes, other.attributes);
	}

	@Override
	public int hashCode() {
		return Objects.hash(plainText, blocks, metadata, tables, images, links, footnotes, attributes);
	}

	@Override
	public String toString() {
		return "DjotContent{" + "plainTextLength=" + plainText.length() + ", blocks=" + blocks.size() + ", tables="
				+ tables.size() + ", images=" + images.size() + ", links=" + links.size() + ", footnotes="
				+ footnotes.size() + ", attributes=" + attributes.size() + '}';
	}

	/**
	 * Represents an element attribute entry (String key to Attributes mapping).
	 */
	public static final class AttributeEntry {
		private final String key;
		private final Attributes attributes;

		@JsonCreator
		public AttributeEntry(@JsonProperty("0") String key, @JsonProperty("1") Attributes attributes) {
			this.key = Objects.requireNonNull(key, "key must not be null");
			this.attributes = Objects.requireNonNull(attributes, "attributes must not be null");
		}

		/**
		 * Creates a new attribute entry.
		 *
		 * @param key
		 *            the element identifier
		 * @param attributes
		 *            the attributes for this element
		 * @return a new AttributeEntry instance
		 */
		public static AttributeEntry of(String key, Attributes attributes) {
			return new AttributeEntry(key, attributes);
		}

		/**
		 * Get the element identifier key.
		 *
		 * @return element key
		 */
		public String getKey() {
			return key;
		}

		/**
		 * Get the attributes for this element.
		 *
		 * @return attributes
		 */
		public Attributes getAttributes() {
			return attributes;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj) {
				return true;
			}
			if (!(obj instanceof AttributeEntry)) {
				return false;
			}
			AttributeEntry other = (AttributeEntry) obj;
			return Objects.equals(key, other.key) && Objects.equals(attributes, other.attributes);
		}

		@Override
		public int hashCode() {
			return Objects.hash(key, attributes);
		}

		@Override
		public String toString() {
			return "AttributeEntry{" + "key='" + key + '\'' + ", attributes=" + attributes + '}';
		}
	}
}

File: kreuzberg/packages/php/src/Types/DjotContent.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Structured Djot document representation.
 *
 * Provides rich, block-level document structure with formatting,
 * tables, images, links, and metadata extracted from source documents.
 *
 * @property-read string $plainText Plain text representation for backward compatibility
 * @property-read array<FormattedBlock> $blocks Structured block-level content
 * @property-read Metadata $metadata Metadata from YAML frontmatter
 * @property-read array<Table> $tables Extracted tables as structured data
 * @property-read array<DjotImage> $images Extracted images with metadata
 * @property-read array<DjotLink> $links Extracted links with URLs and titles
 * @property-read array<Footnote> $footnotes Footnote definitions
 * @property-read array<string, mixed>|null $attributes Attributes mapped by element identifier
 */
readonly class DjotContent
{
    /**
     * @param array<FormattedBlock> $blocks
     * @param array<Table> $tables
     * @param array<DjotImage> $images
     * @param array<DjotLink> $links
     * @param array<Footnote> $footnotes
     * @param array<string, mixed>|null $attributes
     */
    public function __construct(
        public string $plainText,
        public array $blocks,
        public Metadata $metadata,
        public array $tables = [],
        public array $images = [],
        public array $links = [],
        public array $footnotes = [],
        public ?array $attributes = null,
    ) {
    }

    /**
     * Create DjotContent from array returned by extension.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $plainText */
        $plainText = $data['plain_text'] ?? '';

        /** @var array<string, mixed> $metadataData */
        $metadataData = $data['metadata'] ?? [];

        /** @var array<array<string, mixed>> $blocksData */
        $blocksData = $data['blocks'] ?? [];

        $blocks = array_map(
            /** @param array<string, mixed> $block */
            static fn (array $block): FormattedBlock => FormattedBlock::fromArray($block),
            $blocksData,
        );

        /** @var array<array<string, mixed>> $tablesData */
        $tablesData = $data['tables'] ?? [];

        $tables = array_map(
            /** @param array<string, mixed> $table */
            static fn (array $table): Table => Table::fromArray($table),
            $tablesData,
        );

        $images = [];
        if (isset($data['images'])) {
            /** @var array<array<string, mixed>> $imagesData */
            $imagesData = $data['images'];
            $images = array_map(
                /** @param array<string, mixed> $image */
                static fn (array $image): DjotImage => DjotImage::fromArray($image),
                $imagesData,
            );
        }

        $links = [];
        if (isset($data['links'])) {
            /** @var array<array<string, mixed>> $linksData */
            $linksData = $data['links'];
            $links = array_map(
                /** @param array<string, mixed> $link */
                static fn (array $link): DjotLink => DjotLink::fromArray($link),
                $linksData,
            );
        }

        $footnotes = [];
        if (isset($data['footnotes'])) {
            /** @var array<array<string, mixed>> $footnotesData */
            $footnotesData = $data['footnotes'];
            $footnotes = array_map(
                /** @param array<string, mixed> $footnote */
                static fn (array $footnote): Footnote => Footnote::fromArray($footnote),
                $footnotesData,
            );
        }

        /** @var array<string, mixed>|null $attributes */
        $attributes = $data['attributes'] ?? null;

        return new self(
            plainText: $plainText,
            blocks: $blocks,
            metadata: Metadata::fromArray($metadataData),
            tables: $tables,
            images: $images,
            links: $links,
            footnotes: $footnotes,
            attributes: $attributes,
        );
    }
}
